design/community/book-proposal.md

```diff
@@ -0,0 +1,241 @@
+<!--
+This work is licensed under a Creative Commons Attribution 3.0
+Unported License.
+
+http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# metal³ book
+
+## Status
+
+implementable
+
+## Summary
+
+A book for the Metal³ project that describes Baremetal-operator(BMO),
+Cluster-api-provider-metal3(CAPM3), Ip-address-manager(IPAM),
+Hardware-classification-controller(HCC), Metal3-dev-env(Dev-env), etc.
+
+## Motivation
+
+Currently all the documentation describing different parts
+of the Metal³ project is scattered across different repositories
+which doesn’t bring a good user experience from the documentation
+perspective.
+
+### Goals
+
+1. Use asciidoctor toolchain to create the book from the documents
+    available under e.g. /doc in each repository.
+
+2. Migrate existing documents from .`md` to `.adoc` format so that
+    asciidoctor toolchain can render them.
+
+3. Create a final book structure in Metal3-docs and publish the
+    final book from there.
+
+4. Create a partial book structure for each repository.
+
+5. Configure the Netlify for the book preview as part of a PR review.
+
+### Non-Goals
+
+1. Move all the existing documents to a single repository.
+
+2. Have HTML/CSS files within the repository
+
+## Proposal
+
+Use asciidoctor toolchain`[1]` to build a book for the Metal³ project, which
+will combine documents from different repositories such as CAPM3, BMO, HCC,
+IPAM, Dev-env, Metal3-docs and publish/present it as a single book.
+
+### User Stories
+
+#### Story 1
+
+* As the Metal³ developer I would like to have a document where I can find
+    information for the components involved in the Metal³ and the interaction
+    between them whenever I get in trouble.
+
+* As the Metal³ user I would like to have a document that would help me to
+    get started with the project easily.
```

Comment:
I appreciate the effort you're putting into this, and the priority you're giving to the goal of keeping the source files close to the code they are documenting. We went through a similar overhaul of documentation in OpenStack a few years ago, and the lesson I learned there is to carefully consider the audience for everything you write, and publish appropriately. In our case, I worry that having 2 very different audiences introduces enough conflict that squeezing everything into 1 doc will make it harder to consume, and having a complex build system will make it harder to maintain.

What sort of documentation would developers want in this book? Overall architecture? Code structure? Internal APIs? Process instructions (how to submit PRs, etc.)?

What sort of documentation would users want? How to use the tool? External APIs? How to join the community to get/share help?

Those sets of things don't have a lot of overlap, and may be able to take advantage of different tools. Where they do overlap, it's with general, high-level content or end-to-end instructions that may flip back and forth between different components. That will be easier to keep in 1 place.

Here are a few specific suggestions about how to treat the different types of content:

The code structure and internal APIs can be published using https://godoc.org. For example, https://godoc.org/github.com/metal3-io/baremetal-operator/pkg/bmc was generated when I typed our package name into the top level of the site and clicked on a link. It will change as the code changes, without us having to do anything.

Similarly, basic info for external APIs can be generated using a tool like https://doc.crds.dev, although it looks like we have to ask them to index our CRDs before that will work.

More detailed sequential instructions should be written out as prose, of course, but those aren't going to change in fundamental ways all that often, are they? So maybe those could just go into metal3-docs so that we can have nice end-to-end instructions in a form that is easy for users to read and follow. That same book can include suggested deployment architecture details, which can also serve as an overview for developers learning about the project. And since it is all in one place, we can use the same tools that other k8s projects do to build the site from markdown files instead of converting to asciidoctor.

We already have feature-driven architecture documents in metal3-docs as well. I don't think we want those to be part of the "book" but we could link to them where appropriate.

We can put community contact details (mailing lists, slack channels, etc.) on the metal3.io website, so it's all in one place and easy to discover (it should already be there?). Then, we can link to that page from everywhere else that should refer to it.

Reply:
Hi @dhellmann , 
Thank you very much for the detailed comment and sharing your experiences.

I agree that godoc is nice to keep API & code structure related information, but it is limited only to those areas.

What I’m hoping to see in the book is whatever documents we already have under /docs. If you take Baremetal Operator as an example, there are really nice docs that describe for example state machines, or what each field represents in BaremetalHost CR, how to test Baremetal Operator in local k8s clusters, etc.

And the same situation in Cluster-api-provider-metal3. For example, api.md has a description for each field in a couple of different CRs, then how to deploy CAPM3, etc.

It is a bit difficult for me to differentiate the consumers of the book. Because, I would assume that whoever starts contributing for example to BMO, they will at least take a look at those documents before they actually do something with the code. So, why not to put those docs in a book and give it to a contributor or the user. And at the end of a day, isn’t a user == developer, maybe not to Metal3.

If we want to make clear distinctions between the consumers of the book, we could for example make three different sections in the book, first one for the intro, second for the high level picture of the project, and third for the deep information.

In the first section we could give general information. For example, Why Metal3, what does it offer, how to join Metal3, ways of working, where to start, links to mailing lists, zoom, etc. 

```yaml
Introduction
Quick start
How to contribute
Metal3  community
….
```
Of course, some of the information is not documented but I think it is not a big deal to write it down and put it in Metal3-docs and then reference them from the book. I can sign up myself to do that.

In section2 we could continue with a high level guide, which would describe how to get started with Metal3, how to use it, how to deploy CAPM3, BMO, IPAM, HCC, etc. In this part, we would not provide deep information, like what are the custom objects, what does each field represent in them (e.g. api.md).

And last, in section3 we could describe every custom resource, controllers, their interaction, state machines, how BMO interacts with Ironic, etc.

Talking about tooling for the book, If I’m not wrong, many documentations in Kubernetes are presented with Hugo pages. But I’m not sure what is in between that takes  .md files and makes a structure out of it. It could be the Hugo pages, but I’m not sure TBH.
One thing to note is that we don’t want to put complicated mechanisms to generate the book. 
In the proposal it is just a single file that has references to .adoc files in different repositories  and that’s all basically. We don’t have to take care of updating the book - we just need to trigger the build command 1/N times a day. And it will present whatever information we currently have in those referenced .adoc files.

Regarding the asciidoc syntax, it is true that as one of the biggest drawbacks of the asciidoctor we all will need to learn asciidoc syntax but I would say it is not that big a difference compared to markdown. We could put a document in the repo to describe how to do the proper formatting when working with .adoc files and I think that will be good enough to quickly get started writing down in asciidoc format.

Reply:
I can take a look at Hugo pages, and check how can we convert .md files to HTML/CSS.

Reply:
Easy too, select a theme, add it to the repo and put md files there, hugo will run and generate it, similar to pelican or jekyll, but you get a 'website' not something you can easily download.

Example of hugo site: https://openshift.tips/

Reply:
Yes, that's true. The advantage with asciidoctor is that we can have pdf version to download.